ethereum/simplestack/simple_stack.go

```diff
@@ -0,0 +1,167 @@
+package simplestack
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/0xProject/0x-mesh/ethereum/miniheader"
+)
+
+type updateType int
+
+const (
+	Pop updateType = iota
+	Push
+)
+
+// Update represents one update to the stack, either a pop or push of a miniHeader.
+type Update struct {
+	Type       updateType
+	MiniHeader *miniheader.MiniHeader
+}
+
+// SimpleStack is a simple in-memory stack used in tests
+type SimpleStack struct {
+	limit              int
+	miniHeaders        []*miniheader.MiniHeader
+	updates            []*Update
+	mu                 sync.Mutex
+	latestCheckpointID int
+}
+
+// NewSimpleStack instantiates a new SimpleStack
+func NewSimpleStack(retentionLimit int, miniHeaders []*miniheader.MiniHeader) *SimpleStack {
+	return &SimpleStack{
+		limit:       retentionLimit,
+		miniHeaders: miniHeaders,
+		updates:     []*Update{},
+	}
+}
+
+// Peek returns the top of the stack
+func (s *SimpleStack) Peek() (*miniheader.MiniHeader, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if len(s.miniHeaders) == 0 {
+		return nil, nil
+	}
+	return s.miniHeaders[len(s.miniHeaders)-1], nil
+}
+
+// Pop returns the top of the stack and removes it from the stack
+func (s *SimpleStack) Pop() (*miniheader.MiniHeader, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	return s.pop()
+}
+
+func (s *SimpleStack) pop() (*miniheader.MiniHeader, error) {
+	if len(s.miniHeaders) == 0 {
+		return nil, nil
+	}
+	top := s.miniHeaders[len(s.miniHeaders)-1]
+	s.miniHeaders = s.miniHeaders[:len(s.miniHeaders)-1]
+	s.updates = append(s.updates, &Update{
+		Type:       Pop,
+		MiniHeader: top,
+	})
+	return top, nil
+}
+
+// Push adds a miniheader.MiniHeader to the stack
+func (s *SimpleStack) Push(miniHeader *miniheader.MiniHeader) error {
```

Comment:
`push` and `pop` are both called from within `Reset` which already holds a mutex and so we need a way to access the `pop` and `push` functionality without attempting to acquire the mutex we're already holding. I made these methods private since they are only meant to be used internally by other methods that have already acquired a mutex.

Reply:
In that case can you add a comment to `pop` and `push` indicating that the caller should already hold a lock on the mutex?